pkg/imgpkg/bundle/bundle_images_lock.go

```diff
@@ -9,53 +9,83 @@ import (
 	"io"
 	"io/ioutil"
 	"path/filepath"
+	"strings"
 	"sync"
 
+	regname "github.com/google/go-containerregistry/pkg/name"
 	regv1 "github.com/google/go-containerregistry/pkg/v1"
 	"github.com/google/go-containerregistry/pkg/v1/types"
 	"github.com/k14s/imgpkg/pkg/imgpkg/lockconfig"
 	"github.com/k14s/imgpkg/pkg/imgpkg/util"
 )
 
-func (o *Bundle) AllImagesLock(concurrency int) (*ImagesLock, error) {
+func (o *Bundle) AllImagesRefs(concurrency int, logger util.LoggerWithLevels) ([]*Bundle, ImagesRef, error) {
 	throttleReq := util.NewThrottle(concurrency)
-	return o.buildAllImagesLock(&throttleReq, &processedImages{processedImgs: map[string]struct{}{}})
+	bundles, imagesLock, err := o.buildAllImagesLock(&throttleReq, &processedImages{processedImgs: map[string]struct{}{}}, logger)
+	if err != nil {
+		return nil, ImagesRef{}, err
+	}
+
+	imagesRef := imagesLock.ImagesRefs()
+	// Ensure that the correct IsBundle flag is provided.
+	// This loop needs to happen because we skipped some images for some bundle, and only at this point we have
+	// the full list of ImagesRef created and can fill the gaps inside each bundle
+	for _, bundle := range bundles {
+		for _, ref := range bundle.ImagesRef() {
+			imgRef, found := imagesRef.ImageRef(ref.Image)
+			if !found {
+				panic(fmt.Sprintf("Internal inconsistency: The Image '%s' cannot be found in the total list of images", ref.Image))
+			}
+
+			bundle.AddImagesRef(imgRef)
+		}
+	}
```

Comment:
i think we need to figure out a better way to do this. trying to think what are we actually storing the state for...

Reply:
We are storing the state for each bundle of the Images that are part of it. This later is used when we are trying to generate each individual Locations OCI Image to find from the list of the ProcessedImages the ones that each bundle needs.